include/cuco/detail/static_map/static_map_ref.inl

```diff
@@ -549,6 +548,235 @@ class operator_impl<
+   *   have this signature to atomically apply the operation.
+   *
+   * @param group The Cooperative Group used to perform group insert
+   * @param value The element to insert
+   * @param op The callable object to perform binary operation between existing value at the slot
+   *  and the element to insert.
+   */
+
+  template <typename Value, typename Op>
+  __device__ void insert_or_apply(cooperative_groups::thread_block_tile<cg_size> const& group,
+                                  Value const& value,
+                                  Op op)
+  {
+    static_assert(
+      cuda::std::is_invocable_v<Op, cuda::atomic_ref<T, Scope>, T>,
+      "insert_or_apply expects `Op` to be a callable as `Op(cuda::atomic_ref<T, Scope>, T)`");
+
+    ref_type& ref_ = static_cast<ref_type&>(*this);
+
+    auto const val         = ref_.impl_.heterogeneous_value(value);
+    auto const key         = ref_.impl_.extract_key(val);
+    auto& probing_scheme   = ref_.impl_.probing_scheme();
+    auto storage_ref       = ref_.impl_.storage_ref();
+    auto probing_iter      = probing_scheme(group, key, storage_ref.window_extent());
+    auto const empty_value = ref_.impl_.empty_slot_sentinel().second;
+
+    while (true) {
+      auto const window_slots = storage_ref[*probing_iter];
+
+      auto const [state, intra_window_index] = [&]() {
+        auto res = detail::equal_result::UNEQUAL;
+        for (auto i = 0; i < window_size; ++i) {
+          res = ref_.impl_.predicate_.operator()<is_insert::YES>(key, window_slots[i].first);
+          if (res != detail::equal_result::UNEQUAL) {
+            return detail::window_probing_results{res, i};
+          }
+        }
+        // returns dummy index `-1` for UNEQUAL
+        return detail::window_probing_results{res, -1};
+      }();
+
+      auto* slot_ptr = (storage_ref.data() + *probing_iter)->data() + intra_window_index;
+
+      auto const group_contains_equal = group.ballot(state == detail::equal_result::EQUAL);
+      if (group_contains_equal) {
+        auto const src_lane = __ffs(group_contains_equal) - 1;
+        if (group.thread_rank() == src_lane) {
+          op(cuda::atomic_ref<T, Scope>{slot_ptr->second}, val.second);
+        }
+        return;
+      }
+
+      auto const group_contains_available = group.ballot(state == detail::equal_result::AVAILABLE);
+      if (group_contains_available) {
+        auto const src_lane = __ffs(group_contains_available) - 1;
+        auto const status   = [&, target_idx = intra_window_index]() {
+          if (group.thread_rank() != src_lane) { return insert_result::CONTINUE; }
+          return ref_.impl_.attempt_insert_stable(slot_ptr, window_slots[target_idx], val);
+        }();
+
+        switch (group.shfl(status, src_lane)) {
+          case insert_result::SUCCESS: return;
+          case insert_result::DUPLICATE: {
+            if (group.thread_rank() == src_lane) {
+              ref_.impl_.wait_for_payload(slot_ptr->second, empty_value);
```

Comment:
Similar to the scalar version, wait only when necessary (`sizeof(value_type) > 8`)

Reply:
Not in this PR, but I'm wondering if we should move the wait logic to `attempt_insert_stable`.